src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2575,19 +2569,37 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char
             }
 
             outBuffer = output.Slice(cursor);
+            if (!newLine)
+            {
+                WriteConsole(consoleHandle, outBuffer);
+                return;
+            }
+
+            char[] rentedArray = null;
+            string lineEnding = Environment.NewLine;
+            int size = outBuffer.Length + lineEnding.Length;
+
+            // We expect the 'size' will often be small, and thus optimize that case with 'stackalloc'.
+            Span<char> buffer = size <= MaxStackAllocSize ? stackalloc char[size] : default;
```

Comment:
@SeeminglyScience and I did some investigation about using `stackalloc` with constant size VS. variable size, and the conclusion is:

1. By default, there is no much difference when `size = 512`. If size is smaller, then using variable size is way faster than doing `(stackalloc char[512]).Slice(0, size)`.
2. When applying `SkipLocalsInit` attribute to the method, `(stackalloc char[512]).Slice(0, size)` is faster than `stackallow char[size]` by a constant value (`1.5 - 1.7ns` on my machine).

The most expensive part of `stackalloc char[xx]` is from initializing the memory with zero (zeroing memory), which JIT does by default. In this case, using variable size is better than using a constant Max size and then slicing, because the larger the memory chunk is, the longer it takes for zeroing the memory.

When applying `SkipLocalsInit` attribute to the method, you are telling the compiler to not initialize local variables, so `stackalloc` operation will be much faster regardless of using constant size or variable size. However, using variable size will incur additional security checks, which makes using constant size and then slicing more performant (`1.5 - 1.7ns` on my machine).

Here are the benchmark results for refernece:
```c#
    public class SkipLocalsInitBenchmark
    {
        [Params(20, 512)]
        public int Size { get; set; }

        [Benchmark]
        [SkipLocalsInit]  // comment out this for comparison
        public char ConstantSize()
        {
            Span<char> s = (stackalloc char[512]).Slice(0, Size);
            return s[0];
        }

        [Benchmark]
        [SkipLocalsInit]  // comment out this for comparison
        public char VariableSize()
        {
            Span<char> s = stackalloc char[Size];
            return s[0];
        }
    }
```
```
## Results without 'SkipLocalsInit' attribute

|       Method | Size |      Mean |     Error |    StdDev |
|------------- |----- |----------:|----------:|----------:|
| ConstantSize |   20 | 28.625 ns | 0.4742 ns | 0.4435 ns |
| VariableSize |   20 |  2.371 ns | 0.0729 ns | 0.0682 ns |
| ConstantSize |  512 | 28.418 ns | 0.2296 ns | 0.1917 ns |
| VariableSize |  512 | 30.529 ns | 0.4140 ns | 0.3872 ns |
```
```
## Results with 'SkipLocalsInit' attribute

|       Method | Size |      Mean |     Error |    StdDev |
|------------- |----- |----------:|----------:|----------:|
| ConstantSize |   20 | 0.6933 ns | 0.0245 ns | 0.0229 ns |
| VariableSize |   20 | 2.3601 ns | 0.0722 ns | 0.0675 ns |
| ConstantSize |  512 | 0.8934 ns | 0.0333 ns | 0.0295 ns |
| VariableSize |  512 | 2.3836 ns | 0.0717 ns | 0.0670 ns |
```

According to the .NET team, once https://github.com/dotnet/runtime/issues/63500 is addressed in .NET runtime, then `(stackalloc char[512]).Slice(0, size)` will always be faster than `stackalloc char[size]`. But that issue is currently in `Future` milestone.

Back to this PR, the bottleneck of this method is the Win32 API call to `NativeMethods.WriteConsole`, so trying to optimize `stackalloc` is really micro-optimization in my opinion. I will go with using the variable size for simplicity (also, it's very likely faster by default as of today). But it's good to know that when dealing with highly performance sensitive code, `SkipLocalsInit` and `(stackalloc char[ConstSize]).Slice(0, size)` will give you the best performance when it really matters.

Reply:
Thanks for great investigations!

Why do we use `default` and Slice in line 2590? I think [follow ](https://source.dot.net/#System.Console/System/IO/CachedConsoleStream.cs,29)works:
```suggestion
            Span<char> buffer = size <= MaxStackAllocSize ? stackalloc char[size] : (rentedArray = ArrayPool<char>.Shared.Rent(size));
```

Reply:
You need to return the array afterwards, and thus need to save the array to a local variable.

Reply:
Sorry, I don't understand your comment. In my example we save the array to a local variable.

Reply:
Sorry, I misread your comment :)

`ArrayPool.Rent(size)` returns an array of at least `size` length, and more often than not it's bigger than `size`, so we will have to call `Slice` on it before passing to`WriteConsole(handle, buffer)`. Then this needs to be written as
`: (rentedArray = ArrayPool<char>.Shared.Rent(size)).AsSpan().Slice(0, size);`, which is less readable in my opinion.

Reply:
I see your point. Thanks for clarify!

#Closed